tests/mock_network/churn.rs

```diff
@@ -20,65 +20,107 @@ use routing::{
 };
 use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
 
-/// Randomly removes some nodes, but <1/3 from each section and never node 0.
-/// Never trigger merge: never remove enough nodes to drop to `elder_size`.
-/// max_per_pfx: limits dropping to the specified count per pfx. It would also
-/// skip prefixes randomly allowing sections to split if this is executed in the same
-/// iteration as `add_nodes_and_poll`.
+/// Randomly removes some nodes.
+///
+/// Limits the number of nodes simultaneously dropped from a section such that the section still
+/// remains functional (capable of reaching consensus), also accounting for the fact that any
+/// dropped node might trigger relocation of other nodes. This limit is currently possibly too
+/// conservative and might change in the future, but it still allows removing at least one node per
+/// section.
+///
+/// If `max_per_pfx` is set, never drops more than that number of nodes per section.
 ///
 /// Note: it's necessary to call `poll_all` afterwards, as this function doesn't call it itself.
 fn drop_random_nodes<R: Rng>(
     rng: &mut R,
     nodes: &mut Vec<TestNode>,
     max_per_pfx: Option<usize>,
 ) -> BTreeSet<XorName> {
-    let mut dropped_nodes = BTreeSet::new();
-    let elder_size = |node: &TestNode| unwrap!(node.inner.elder_size());
-    let node_section_size = |node: &TestNode| {
-        node.inner
-            .section_elders(unwrap!(node.inner.our_prefix(), "{}", node.inner))
-            .len()
+    #[derive(Default)]
+    struct SectionInfo {
+        initial_elder_count: usize,
+        initial_other_count: usize,
+        dropped_elder_count: usize,
+        dropped_other_count: usize,
     };
-    let sections: BTreeMap<_, _> = nodes
-        .iter()
-        .map(|node| {
-            let initial_size = node_section_size(node);
-            let min_size = elder_size(node);
-            let max_drop = initial_size.saturating_sub(min_size);
-            (*node.our_prefix(), (initial_size, max_drop))
-        })
-        .collect();
-    let mut drop_count: BTreeMap<_, _> = sections.keys().map(|pfx| (*pfx, 0)).collect();
-    loop {
-        let i = gen_range(rng, 1, nodes.len());
-        let pfx = nodes[i].our_prefix();
-        let (initial_size, max_drop) = sections[&pfx];
-        if drop_count.is_empty() {
-            break;
-        } else if drop_count.get(&pfx).is_none() {
+
+    // 10% probability that a node will be dropped.
+    let drop_probability = 0.1;
+
+    let mut sections = HashMap::<_, SectionInfo>::new();
```

Comment:
Maybe worth a sub function named since I had to look in depth to understand what was going on.
maybe:
`fn count_elders_and_others_nodes_by_prefix() -> SectionCountInfo`